server/src/main/java/org/opensearch/gateway/ReplicaShardBatchAllocator.java

```diff
@@ -135,34 +158,54 @@ public HashMap<ShardRouting, AllocateUnassignedDecision> makeAllocationDecision(
         final FetchResult<NodeStoreFilesMetadataBatch> shardsState = fetchData(eligibleShards, ineligibleShards, allocation);
 
-        for (ShardRouting unassignedShard : eligibleShards) {
-            Tuple<Decision, Map<String, NodeAllocationResult>> result = nodeAllocationDecisions.get(unassignedShard);
-            shardAllocationDecisions.put(
-                unassignedShard,
-                getAllocationDecision(
-                    unassignedShard,
-                    allocation,
-                    convertToNodeStoreFilesMetadataMap(unassignedShard, shardsState),
-                    result,
-                    logger
-                )
-            );
+        List<ShardId> shardIdsFromBatch = shardRoutings.stream().map(shardRouting -> shardRouting.shardId()).collect(Collectors.toList());
+        RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
+        while (iterator.hasNext()) {
+            ShardRouting unassignedShard = iterator.next();
+            // There will be only one entry for the shard in the unassigned shards batch
+            // for a shard with multiple unassigned replicas, hence we are comparing the shard ids
+            // instead of ShardRouting in-order to evaluate shard assignment for all unassigned replicas of a shard.
+            if (!unassignedShard.primary() && shardIdsFromBatch.contains(unassignedShard.shardId())) {
+                AllocateUnassignedDecision allocateUnassignedDecision;
+                if (ineligibleShardAllocationDecisions.containsKey(unassignedShard)) {
+                    allocateUnassignedDecision = ineligibleShardAllocationDecisions.get(unassignedShard);
+                } else {
+                    if (!isResponsibleFor(unassignedShard)) {
+                        allocateUnassignedDecision = AllocateUnassignedDecision.NOT_TAKEN;
+                    } else {
+                        Tuple<Decision, Map<String, NodeAllocationResult>> result = canBeAllocatedToAtLeastOneNode(
+                            unassignedShard,
+                            allocation
+                        );
+                        Decision allocationDecision = result.v1();
+                        if (allocationDecision.type() != Decision.Type.YES && (!explain || !hasInitiatedFetching(unassignedShard))) {
+                            // only return early if we are not in explain mode, or we are in explain mode but we have not
+                            // yet attempted to fetch any shard data
+                            logger.trace("{}: ignoring allocation, can't be allocated on any node", unassignedShard);
+                            allocateUnassignedDecision = AllocateUnassignedDecision.no(
+                                UnassignedInfo.AllocationStatus.fromDecision(allocationDecision.type()),
+                                result.v2() != null ? new ArrayList<>(result.v2().values()) : null
+                            );
```

Comment:
Its a bit difficult to extract this common block to a separate method because this code block along with making ineligibility decision is also computing the value for ```Tuple<Decision, Map<String, NodeAllocationResult>> result``` ,  a reference to the result has to be maintained to get the final allocation decision.

If we still want to refactor we can pass the reference to the result variable from caller method to the common method and assign this variable method inside it. This approach though doesn't look cleaner to me.  What do you think?

Reply:
Refactored.